crates/sui-network/src/state_sync.rs

```diff
@@ -0,0 +1,1433 @@
+            sequence_number_to_digest: HashMap::new(),
+        }
+        .pipe(RwLock::new)
+        .pipe(Arc::new);
+
+        let server = Server {
+            store: store.clone(),
+            peer_heights: peer_heights.clone(),
+            tx: weak_sender,
+        };
+
+        (
+            UnstartedStateSync {
+                handle,
+                mailbox,
+                store,
+                peer_heights,
+                checkpoint_event_sender,
+            },
+            server,
+        )
+    }
+}
+
+pub struct UnstartedStateSync<S> {
+    handle: Handle,
+    mailbox: mpsc::Receiver<StateSyncMessage>,
+    store: S,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    checkpoint_event_sender: broadcast::Sender<VerifiedCheckpoint>,
+}
+
+impl<S> UnstartedStateSync<S>
+where
+    S: CheckpointStore + Clone + Send + Sync + 'static,
+{
+    fn build(self, network: anemo::Network) -> (StateSyncEventLoop<S>, Handle) {
+        let Self {
+            handle,
+            mailbox,
+            store,
+            peer_heights,
+            checkpoint_event_sender,
+        } = self;
+
+        (
+            StateSyncEventLoop {
+                mailbox,
+                weak_sender: handle.sender.downgrade(),
+                tasks: JoinSet::new(),
+                sync_tasks: JoinSet::new(),
+                store,
+                peer_heights,
+                checkpoint_event_sender,
+                network,
+            },
+            handle,
+        )
+    }
+
+    pub fn start(self, network: anemo::Network) -> Handle {
+        let (event_loop, handle) = self.build(network);
+        tokio::spawn(event_loop.start());
+
+        handle
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Handle {
+    sender: mpsc::Sender<StateSyncMessage>,
+    checkpoint_event_sender: broadcast::Sender<VerifiedCheckpoint>,
+}
+
+impl Handle {
+    pub async fn send_checkpoint(&self, checkpoint: VerifiedCheckpoint) {
+        self.sender
+            .send(StateSyncMessage::VerifiedCheckpoint(Box::new(checkpoint)))
+            .await
+            .unwrap()
+    }
+
+    pub fn subscribe_to_checkpoints(&self) -> broadcast::Receiver<VerifiedCheckpoint> {
+        self.checkpoint_event_sender.subscribe()
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub enum GetCheckpointSummaryRequest {
+    Latest,
+    ByDigest(CheckpointDigest),
+    BySequenceNumber(CheckpointSequenceNumber),
+}
+
+#[derive(Debug, Default)]
+struct InMemoryCheckpointStore {
+    highest_processed_checkpoint: Option<(CheckpointSequenceNumber, CheckpointDigest)>,
+    checkpoints: HashMap<CheckpointDigest, VerifiedCheckpoint>,
+    checkpoint_contents: HashMap<CheckpointContentsDigest, CheckpointContents>,
+    sequence_number_to_digest: HashMap<CheckpointSequenceNumber, CheckpointDigest>,
+
+    epoch_to_committee: Vec<Committee>,
+}
+
+impl InMemoryCheckpointStore {
+    pub fn get_checkpoint_by_digest(
```

Comment:
What are your thoughts on the `get_` prefix as in `get_foo` vs `foo`? I think a lot of our code doesn't use the `get`, for example all of the stuff in the narwhal `Committee` type comes to mind (or methods on `anemo` types).

I find myself asking a lot of style questions like this. Do we have a Mysten Rust style guide? Should we? 🤔

Reply:
So generally i'd say we should avoid the get_ prefix on struct member field accessors. I have been using the get_ prefix on things which are not field accessors but rather some more expensive operation, querying storage, a hashmap, or some network request.